### PR TITLE
AFL-fuzz hang/crash fixes

### PR DESCRIPTION
**Pull request by dezgeg**
Fix three issues AFL found in ~30 hours.

### CODE REVIEW

**Comment by copumpkin** [comment]
Nice! Have any information on your AFL setup for Nix? Might also be fun to try on macOS due to the different codepaths we exercise.

**Comment by dtzWill** [comment]
This is great!

Next stop.... [OSS-Fuzz](https://github.com/google/oss-fuzz)?

File: src/libutil/util.cc [hunk 1]
```diff
@@ -1216,7 +1216,7 @@ std::string filterANSIEscapes(const std::string & s, unsigned int width)
 
         else if (*i == '\r')
             // do nothing for now
-            ;
+            i++;
```
Comment (shlevy) [inline src/libutil/util.cc]: Nice, would love a C++ unit testing framework to put this in.

**Comment by edolstra** [comment]
The fix for an infinite loop in replaceString caused an infinite loop: e2d71bd1862cdda65f957c6a6c41016258d9e003.

**Comment by dezgeg** [comment]
Eep. Will take a look.

**Comment by dezgeg** [comment]
@copumpkin The setup is at https://github.com/dezgeg/fuzz-nix. However I don't actually try to build the (potential) generated derivations, just evaluate them with `nix-instantiate --dry-run --read-only-mode`.

**Pull request merged**